core/src/cluster_slots.rs

```diff
@@ -18,34 +22,56 @@ pub type SlotPubkeys = HashMap<Pubkey, u64>;
 #[derive(Default)]
 pub struct ClusterSlots {
     cluster_slots: RwLock<BTreeMap<Slot, Arc<RwLock<SlotPubkeys>>>>,
-    since: RwLock<Option<u64>>,
+    since: AtomicU64,
     validator_stakes: RwLock<Arc<NodeIdToVoteAccounts>>,
     epoch: RwLock<Option<u64>>,
-    self_id: RwLock<Pubkey>,
 }
 
 impl ClusterSlots {
     pub fn lookup(&self, slot: Slot) -> Option<Arc<RwLock<SlotPubkeys>>> {
         self.cluster_slots.read().unwrap().get(&slot).cloned()
     }
     pub fn update(&self, root: Slot, cluster_info: &ClusterInfo, bank_forks: &RwLock<BankForks>) {
-        self.update_peers(cluster_info, bank_forks);
-        let since = *self.since.read().unwrap();
+        self.update_peers(bank_forks);
+        let since = self.since.load(Ordering::Relaxed);
         let (epoch_slots, since) = cluster_info.get_epoch_slots_since(since);
         self.update_internal(root, epoch_slots, since);
     }
     fn update_internal(&self, root: Slot, epoch_slots_list: Vec<EpochSlots>, since: Option<u64>) {
-        for epoch_slots in epoch_slots_list {
-            let slots = epoch_slots.to_slots(root);
-            for slot in &slots {
-                if *slot <= root {
-                    continue;
-                }
-                self.insert_node_id(*slot, epoch_slots.from);
-            }
-        }
+        // Attach validator's total stake.
+        let epoch_slots_list: Vec<_> = {
+            let validator_stakes = self.validator_stakes.read().unwrap();
+            epoch_slots_list
+                .into_iter()
+                .map(|epoch_slots| {
+                    let stake = match validator_stakes.get(&epoch_slots.from) {
+                        Some(v) => v.total_stake,
+                        None => 0,
+                    };
+                    (epoch_slots, stake)
+                })
+                .collect()
+        };
+        let slot_node_stakes = epoch_slots_list
+            .into_iter()
+            .flat_map(|(epoch_slots, stake)| {
+                epoch_slots
+                    .to_slots(root)
+                    .into_iter()
+                    .filter(|slot| *slot > root)
+                    .zip(std::iter::repeat((epoch_slots.from, stake)))
+            })
+            .into_group_map();
         {
             let mut cluster_slots = self.cluster_slots.write().unwrap();
+            for (slot, node_stakes) in slot_node_stakes {
```

Comment:
Hmm, so I'm worried that if there are a lot of updates here, holding the write lock may block repair which is also consuming this `self.cluster_slots`. 

Maybe we can try:
1)  Grabbing the read lock for the `cluster_slots`, cloning out the existing `Arc<RwLock<SlotPubkeys>>` and modifying those.
2) Only grab the write lock on `cluster_slots` to create + clone the missing set of `slot` keys, drop the lock. Modify the cloned `Arc<RwLock<SlotPubkeys>>`

Reply:
makes sense. updated the code.